sip/sip-002-smart-contract-language.md

```diff
@@ -304,28 +304,51 @@ directly, it _can_ read data stored in those smart contracts' maps.
 contracting language. All data in the smart contracts is inherently
 public, and will be readable through querying the underlying database
 in any case.) In order to do so, a contract may use the
-`(fetch-contract-entry)` function, which behaves identically to
-`(fetch-entry)`, though it accepts a contract principal as an argument
+`(contract-map-get)` function, which behaves identically to
+`(map-get)`, though it accepts a contract principal as an argument
 in addition to the map name:
 
 ```
-(fetch-contract-entry
+(contract-map-get
   'contract-principal
   'map-name
   'key-tuple) -> value tuple or none
 
 Example:
 
-(fetch-contract-entry
+(contract-map-get
  'SC3H92H297DX3YDPFHZGH90G8Z4NPH4VE8E83YWAQ
  'name-map
  12234) -> returns owner principal of name represent by integer 12234
 ```
 
-Just as with the `(contract-call)` function, the map name and contract
+Just as with the `(contract-call!)` function, the map name and contract
 principal arguments must be constants, specified at the time of
 publishing.
 
+### Time-shifted Evaluations
+
+The Stacks language supports _historical_ data queries using the
+`(at-block)` function:
+
+```
+(at-block 0x0101010101010101010101010101010101010101010101010101010101010101
```

Comment:
I thought `at-block` was going to take a block height?

Reply:
Actually, what would be most amenable to the way the system reasons about the chain organization would be to have `at-block` take an _offset_ from the chain tip.  For example, `(at-block -5` is "five blocks ago".

Reply:
"five blocks ago" seems like a _very_ fragile way to specify a block -- a miner choosing not to include your contract would radically alter it's behavior.

Using block height is probably a simpler interface, but implementing a mapping from block-height to block hash would need to be done. I'm not radically opposed to it, but using block hash seems to me to offer the most specificity to developers. They will also be able to obtain block hashes from block height using the `get-block-info` method.

Reply:
I generally agree with using a specific block hash, but we'll need to update the MARF fork table so it can efficiently determine whether or not a block is an ancestor of another block, given just their hashes.